plugins/newsql/newsql.c

```diff
@@ -448,6 +455,53 @@ static int get_col_type(struct sqlclntstate *clnt, sqlite3_stmt *stmt, int col)
     return type;
 }
 
+static inline void reset_protobuf_offset(struct sqlclntstate *clnt)
+{
+        ((struct newsql_appdata *)clnt->appdata)->protobuf_offset = 0;
+}
+
+static void *newsql_protobuf_alloc(void *allocator_data, size_t size)
+{
+    struct newsql_appdata *appdata = allocator_data;
+    void *p = NULL;
+    if (size <= appdata->protobuf_size - appdata->protobuf_offset) {
+        p = appdata->protobuf_data + appdata->protobuf_offset;
+        appdata->protobuf_offset += size;
+    } else {
+        p = malloc(size);
+    }
+    return p;
+}
+
+static void newsql_protobuf_free(void *allocator_data, void *p)
+{
+    struct newsql_appdata *appdata = allocator_data;
+    if (p < appdata->protobuf_data || p > (appdata->protobuf_data + appdata->protobuf_size)) {
+        free(p);
+    }
+}
+
+static inline void newsql_protobuf_init(struct newsql_appdata *appdata)
+{
+    appdata->protobuf_size = gbl_protobuf_prealloc_buffer_size;
+    appdata->protobuf_offset = 0;
+    appdata->protobuf_data = malloc(appdata->protobuf_size);
+    appdata->protobuf_allocator.alloc = &newsql_protobuf_alloc;
+#if free == comdb2_free
```

Comment:
Want to understand why this was needed.

Reply:
Agree, I don't like this. You can define new allocator routines in `pb_alloc.h` and use it here. This is already in use in newsql.